### ISSUE TITLE
Does the backend handle Federated learning asynchronously?

### ISSUE THREAD

**Issue by 111Kaushal** [facebookresearch/FLSim#47]
I found this repo from this blog: - https://ai.facebook.com/blog/asynchronous-federated-learning/
However I do not find any mentioning on this repo and also I cannot decipher from the code examples whether this is synchronous version or asynchronous version of Federated learning?
Can you please clarify this for me? And also if this is the asynchronous version how can I dive deeper in to the libraries and look at the code of implementation for the asynch handling mechanism?

Thank you

**Comment by JohnlNguyen** [comment]
Hi, we have the example config to run asyncFL here 
https://github.com/facebookresearch/FLSim/blob/main/examples/configs/sent140_fedbuff_config.json

and the code for async is here
https://github.com/facebookresearch/FLSim/blob/main/flsim/trainers/async_trainer.py

**Comment by 111Kaushal** [comment]
Hello!
Thank you for the reply. I have one more question. I have been trying to
understand the algorithm in the paper. So please correct me if I am wrong
and I kindly request to break it down in much simpler words and terms.
Here's what I understood (ignoring the differential privacy and SecAgg):

1. Every client will train for Q epochs. After that the difference between
the weights of the previously averaged model (yo=w) and this new locally
trained model (yo-yq) is submitted to the cloud. This is now called as
delta or client update. (∆i)
2. Once the server gets K=10 such client updates it then simply adds (∆ t ←
∆ t + ∆i) and divides it by K (∆ t ← ∆ t /K) meaning that it has taken the
average of all the client updates
3. After this point I am a bit lost.  This expression does not make sense
to me : (w t+1 ← w t − ηg∆ t) so you multiply the averaged client update
with the server learning rate which is fine but then take the difference of
the global model weights and  ηg∆ t ??? And now these are the new weights
of the global model?? Shouldn't it be just addition instead of
subtraction?? Cause learning rate is a positive value unless the averaged
delta or averaged client update is negative  , then in that case
negative*negative makes it positive ?? Can you please clarify and please do
correct me if I am going wrong in understanding the overall algorithm. I
really want to understand this in depth.

Thank you
Regards
Kaushal Rathi
